File: lib/features/shared/domain/entities/enums.dart
// Pure domain enums. Each value carries its wire token (the JSON string used by
// the API) and a static `fromWire` that maps an unrecognised value to `unknown`.
// No JSON, Dio or Flutter imports — this file is safe for the domain layer.

enum SeasonStatus {
  upcoming('upcoming'),
  active('active'),
  completed('completed'),
  unknown('unknown');

  const SeasonStatus(this.wire);
  final String wire;

  static SeasonStatus fromWire(String? wire) =>
      values.firstWhere((v) => v.wire == wire, orElse: () => unknown);
}

enum WeekendFormat {
  standard('standard'),
  sprint('sprint'),
  unknown('unknown');

  const WeekendFormat(this.wire);
  final String wire;

  static WeekendFormat fromWire(String? wire) =>
      values.firstWhere((v) => v.wire == wire, orElse: () => unknown);
}

enum EventStatus {
  scheduled('scheduled'),
  upcoming('upcoming'),
  inProgress('in_progress'),
  completed('completed'),
  postponed('postponed'),
  cancelled('cancelled'),
  unknown('unknown');

  const EventStatus(this.wire);
  final String wire;

  static EventStatus fromWire(String? wire) =>
      values.firstWhere((v) => v.wire == wire, orElse: () => unknown);
}

enum SessionType {
  practice1('practice_1'),
  practice2('practice_2'),
  practice3('practice_3'),
  sprintQualifying('sprint_qualifying'),
  sprint('sprint'),
  qualifying('qualifying'),
  race('race'),
  unknown('unknown');

  const SessionType(this.wire);
  final String wire;

  static SessionType fromWire(String? wire) =>
      values.firstWhere((v) => v.wire == wire, orElse: () => unknown);
}

enum SessionStatus {
  scheduled('scheduled'),
  live('live'),
  completed('completed'),
  cancelled('cancelled'),
  postponed('postponed'),
  unknown('unknown');

  const SessionStatus(this.wire);
  final String wire;

  static SessionStatus fromWire(String? wire) =>
      values.firstWhere((v) => v.wire == wire, orElse: () => unknown);
}

enum DriverRole {
  race('race'),
  reserve('reserve'),
  test('test'),
  unknown('unknown');

  const DriverRole(this.wire);
  final String wire;

  static DriverRole fromWire(String? wire) =>
      values.firstWhere((v) => v.wire == wire, orElse: () => unknown);
}

enum ResultStatus {
  provisional('provisional'),
  finalResult('final'),
  unavailable('unavailable'),
  unknown('unknown');

  const ResultStatus(this.wire);
  final String wire;

  static ResultStatus fromWire(String? wire) =>
      values.firstWhere((v) => v.wire == wire, orElse: () => unknown);
}

enum FinishStatus {
  finished('finished'),
  lapped('lapped'),
  dnf('dnf'),
  dns('dns'),
  dsq('dsq'),
  dnq('dnq'),
  unknown('unknown');

  const FinishStatus(this.wire);
  final String wire;

  static FinishStatus fromWire(String? wire) =>
      values.firstWhere((v) => v.wire == wire, orElse: () => unknown);
}

enum CircuitDirection {
  clockwise('clockwise'),
  counterClockwise('counter_clockwise'),
  unknown('unknown');

  const CircuitDirection(this.wire);
  final String wire;

  static CircuitDirection fromWire(String? wire) =>
      values.firstWhere((v) => v.wire == wire, orElse: () => unknown);
}

enum MediaEntityType {
  driver('driver'),
  constructor('constructor'),
  circuit('circuit'),
  grandPrix('grand_prix'),
  placeholder('placeholder'),
  unknown('unknown');

  const MediaEntityType(this.wire);
  final String wire;

  static MediaEntityType fromWire(String? wire) =>
      values.firstWhere((v) => v.wire == wire, orElse: () => unknown);
}

enum MediaCategory {
  portrait('portrait'),
  logo('logo'),
  car('car'),
  circuitLayout('circuit_layout'),
  hero('hero'),
  thumbnail('thumbnail'),
  unknown('unknown');

  const MediaCategory(this.wire);
  final String wire;

  static MediaCategory fromWire(String? wire) =>
      values.firstWhere((v) => v.wire == wire, orElse: () => unknown);
}

enum MediaFormat {
  webp('webp'),
  avif('avif'),
  png('png'),
  jpeg('jpeg'),
  unknown('unknown');

  const MediaFormat(this.wire);
  final String wire;

  static MediaFormat fromWire(String? wire) =>
      values.firstWhere((v) => v.wire == wire, orElse: () => unknown);
}

File: services/edge-api/src/contract/enums.ts
// Public contract enums, expressed as string-literal unions. Every enum carries
// an `unknown` member so an unrecognised value maps to `unknown` rather than
// throwing. Mirrors docs/technical/GridView_Domain_Model.md section 5.4.

export const SEASON_STATUSES = [
  'upcoming',
  'active',
  'completed',
  'unknown',
] as const;
export type SeasonStatus = (typeof SEASON_STATUSES)[number];

export const WEEKEND_FORMATS = ['standard', 'sprint', 'unknown'] as const;
export type WeekendFormat = (typeof WEEKEND_FORMATS)[number];

export const EVENT_STATUSES = [
  'scheduled',
  'upcoming',
  'in_progress',
  'completed',
  'postponed',
  'cancelled',
  'unknown',
] as const;
export type EventStatus = (typeof EVENT_STATUSES)[number];

export const SESSION_TYPES = [
  'practice_1',
  'practice_2',
  'practice_3',
  'sprint_qualifying',
  'sprint',
  'qualifying',
  'race',
  'unknown',
] as const;
export type SessionType = (typeof SESSION_TYPES)[number];

export const SESSION_STATUSES = [
  'scheduled',
  'live',
  'completed',
  'cancelled',
  'postponed',
  'unknown',
] as const;
export type SessionStatus = (typeof SESSION_STATUSES)[number];

export const DRIVER_ROLES = ['race', 'reserve', 'test', 'unknown'] as const;
export type DriverRole = (typeof DRIVER_ROLES)[number];

export const RESULT_STATUSES = [
  'provisional',
  'final',
  'unavailable',
  'unknown',
] as const;
export type ResultStatus = (typeof RESULT_STATUSES)[number];

export const FINISH_STATUSES = [
  'finished',
  'lapped',
  'dnf',
  'dns',
  'dsq',
  'dnq',
  'unknown',
] as const;
export type FinishStatus = (typeof FINISH_STATUSES)[number];

export const CIRCUIT_DIRECTIONS = [
  'clockwise',
  'counter_clockwise',
  'unknown',
] as const;
export type CircuitDirection = (typeof CIRCUIT_DIRECTIONS)[number];

export const MEDIA_ENTITY_TYPES = [
  'driver',
  'constructor',
  'circuit',
  'grand_prix',
  'placeholder',
  'unknown',
] as const;
export type MediaEntityType = (typeof MEDIA_ENTITY_TYPES)[number];

export const MEDIA_CATEGORIES = [
  'portrait',
  'logo',
  'car',
  'circuit_layout',
  'hero',
  'thumbnail',
  'unknown',
] as const;
export type MediaCategory = (typeof MEDIA_CATEGORIES)[number];

export const MEDIA_FORMATS = [
  'webp',
  'avif',
  'png',
  'jpeg',
  'unknown',
] as const;
export type MediaFormat = (typeof MEDIA_FORMATS)[number];

export const MEDIA_VARIANT_NAMES = [
  'thumbnail',
  'card',
  'detail',
  'hero',
] as const;
export type MediaVariantName = (typeof MEDIA_VARIANT_NAMES)[number];

/**
 * Maps a raw value to a member of `allowed`, falling back to `fallback`
 * (conventionally `unknown`) when the value is not recognised. Additive-safe.
 */
export function parseEnum<T extends string>(
  value: unknown,
  allowed: readonly T[],
  fallback: T,
): T {
  return typeof value === 'string' &&
    (allowed as readonly string[]).includes(value)
    ? (value as T)
    : fallback;
}
